src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityAnalyzer.ImmutableTypeParameterArgumentAnalysis.cs

```diff
@@ -0,0 +1,642 @@
+		) {
+			INamedTypeSymbol? namedType = GetNamedTypeRecursive( type );
+			if( namedType is null ) {
+				return;
+			}
+
+			if( namedType.ContainingType is not null ) {
+				AnalyzeTypeRecursive(
+					reportDiagnostic,
+					annotationsContext,
+					immutabilityContext,
+					namedType.ContainingType,
+					SelectLeftSyntax( getAnalyzedSyntax )
+				);
+			}
+
+			AnalyzeTypeArgumentsRecursive(
+				reportDiagnostic,
+				annotationsContext,
+				immutabilityContext,
+				namedType.TypeArguments,
+				namedType.TypeParameters,
+				getAnalyzedSyntax
+			);
+
+			static INamedTypeSymbol? GetNamedTypeRecursive( ITypeSymbol type ) => type switch {
+				INamedTypeSymbol namedType => namedType,
+				IArrayTypeSymbol arrayType => GetNamedTypeRecursive( arrayType.ElementType ),
+				_ => null
+			};
+		}
+
+		private static Func<SyntaxNodeOrToken> SelectLeftSyntax( Func<SyntaxNodeOrToken> getSyntax ) =>
+			() => SelectLeftSyntax( getSyntax() );
+
+		private static SyntaxNodeOrToken SelectLeftSyntax( SyntaxNodeOrToken syntax ) {
+			if( syntax.IsToken ) {
+				return syntax;
+			}
+
+			return syntax.AsNode() switch {
+				MemberAccessExpressionSyntax memberAccess => memberAccess.Expression,
+				QualifiedNameSyntax qualifiedName => qualifiedName.Left,
+				_ => syntax,
+			};
+		}
+
+		private static void AnalyzeTypeArgumentsRecursive(
+			Action<Diagnostic> reportDiagnostic,
+			AnnotationsContext annotationsContext,
+			ImmutabilityContext immutabilityContext,
+			ImmutableArray<ITypeSymbol> typeArguments,
+			ImmutableArray<ITypeParameterSymbol> typeParameters,
+			Func<SyntaxNodeOrToken> getAnalyzedSyntax
+		) {
+			getAnalyzedSyntax = SelectRightSyntax( getAnalyzedSyntax );
+
+			for( var i = 0; i < typeArguments.Length; ++i ) {
+				ITypeSymbol argument = typeArguments[ i ];
+				ITypeParameterSymbol typeParameter = typeParameters[ i ];
+
+				SyntaxNodeOrToken getThisArgument() => GetTypeArgumentSyntax( getAnalyzedSyntax(), i );
+
+				AnalyzeTypeRecursive(
+					reportDiagnostic,
+					annotationsContext,
+					immutabilityContext,
+					argument,
+					getThisArgument
+				);
+
+				if( !annotationsContext.Objects.Immutable.IsDefined( typeParameter ) ) {
+					continue;
+				}
+
+				if( !immutabilityContext.IsImmutable(
+					new ImmutabilityQuery(
+						ImmutableTypeKind.Total,
+						argument
+					),
+					getLocation: () => getThisArgument().GetLocation(),
+					out Diagnostic diagnostic
+				) ) {
+					// TODO: not necessarily a good diagnostic for this use-case
+					reportDiagnostic( diagnostic );
+				}
+			}
+		}
+
+		private static Func<SyntaxNodeOrToken> SelectRightSyntax( Func<SyntaxNodeOrToken> getSyntax )
+			=> () => SelectRightSyntax( getSyntax() );
+
+		private static SyntaxNodeOrToken SelectRightSyntax( SyntaxNodeOrToken syntax ) {
+			if( syntax.IsToken ) {
+				return syntax;
+			}
+
+			return syntax.AsNode() switch {
+				ArrayTypeSyntax arrayType => arrayType.ElementType,

```

Comment:
Somewhat interesting - given `int[][]`, `IArrayTypeSymbol.ElementType` is `int[]`, but for syntax, `ElementType` is `int` with a list of 2 rank specifiers.